tests/Reduce/ToFirstAndLastTest.php

```diff
@@ -0,0 +1,273 @@
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider dataProviderForGenerators
+     * @param        \Generator $data
+     * @param        mixed $expected
+     */
+    public function testGenerators(\Generator $data, $expected)
+    {
+        // When
+        $result = Reduce::toFirstAndLast($data);
+
+        // Then
+        $this->assertSame($expected, $result);
+    }
+
+    public function dataProviderForGenerators(): array
+    {
+        $gen = fn (array $data) => GeneratorFixture::getGenerator($data);
+
+        return [
+            [
+                $gen([0]),
+                [0, 0],
+            ],
+            [
+                $gen([null]),
+                [null, null],
+            ],
+            [
+                $gen(['']),
+                ['', ''],
+            ],
+            [
+                $gen(['', null]),
+                ['', null],
+            ],
+            [
+                $gen([3, 2]),
+                [3, 2],
+            ],
+            [
+                $gen([1, 2, 3]),
+                [1, 3],
+            ],
+            [
+                $gen([1.1, 1.1, 2.1, 2.1, 3.1, 3.1]),
+                [1.1, 3.1],
+            ],
+            [
+                $gen([[1], '2', 3]),
+                [[1], 3],
+            ],
+            [
+                $gen([false, [1], '2', 3]),
+                [false, 3],
+            ],
+            [
+                $gen([true, [1], '2', 3]),
+                [true, 3],
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider dataProviderForIterators
+     * @param        \Iterator $data
+     * @param        mixed $expected
+     */
+    public function testIterators(\Iterator $data, $expected)
+    {
+        // When
+        $result = Reduce::toFirstAndLast($data);
+
+        // Then
+        $this->assertSame($expected, $result);
+    }
+
+    public function dataProviderForIterators(): array
+    {
+        $iter = fn (array $data) => new ArrayIteratorFixture($data);
+
+        return [
+            [
+                $iter([0]),
+                [0, 0],
+            ],
+            [
+                $iter([null]),
+                [null, null],
+            ],
+            [
+                $iter(['']),
+                ['', ''],
+            ],
+            [
+                $iter(['', null]),
+                ['', null],
+            ],
+            [
+                $iter([3, 2]),
+                [3, 2],
+            ],
+            [
+                $iter([1, 2, 3]),
+                [1, 3],
+            ],
+            [
+                $iter([1.1, 1.1, 2.1, 2.1, 3.1, 3.1]),
+                [1.1, 3.1],
+            ],
+            [
+                $iter([[1], '2', 3]),
+                [[1], 3],
+            ],
+            [
+                $iter([false, [1], '2', 3]),
+                [false, 3],
+            ],
+            [
+                $iter([true, [1], '2', 3]),
+                [true, 3],
+            ],
```

Comment:
Suggestion: Add a test with single value contained in a `NoRewindIterator`. Something like this:

```php
[
    new \NoRewindIterator(new \ArrayIterator([2])),
    [2, 2]
],
```

Due to how you are breaking out of the `foreach` loop in the `toFirst` method this just happens to work due to an implementation detail of foreach loops. This test will make sure this continues to work if the behavior is refactored. Implemented differently this test might fail.

Reply:
Done.